### PR TITLE
Merge

### PR DESCRIPTION
**Pull request by lums658**
This PR supersedes #149 -- it comprises `main` with a squash merge of #149.  The test below is from the original PR.

This PR includes the implementation of two graph-based indexing algorithms: nn-descent and vamana.  The cpp_api branch was also merged into this one, so there is also code in this PR from that branch.

CPP API
======
For C++ library API
* api.h - Type-erased interface between Python and C++, including `Vector`, `FeatureVector`, and `Index`.  The type-erased classes take a URI and read in the corresponding TileDB array.  The underlying type of the `Vector`, `FeatureVector`, or `Index` class is determined by the data type of the TileDB array.
* concepts.h - Initial introduction of C++20 concepts.  A collection of concepts for defining the API to the vector search library.  This allows us to abstract away the format of the data storage (e.g., we don't need to worry about row-major or column-major).
* Use concepts for template parameters in function templates.  Replace type-specific calls in the library (e.g., in query functions, matrix classes) with CPOs.  There will be many small diffs due to this replacement.  The two most important concepts are
  * A `feature_vector` is a contiguous `random_access_range` with the addition of `operator[]` on the container. Moreover, the expression `dimension(v)` must be valid for a type to meet the requirements of `feature_vector` (which can be implemented with a customization point object, or CPO).
  * A `feature_vector_range` is a container of `feature_vector`. The ith `feature_vector` is obtained via `operator[]`. The expressions `dimension(r)` and `num_vectors(r)` must be valid.  `feature_vector_range` is not actually a range (in the sense of `std::ranges`, since we don't provide the required iterators) so we will need to change the name at some point. Or we could consider requiring `feature_vector_range` actually meet the requirements of `std::range` -- in which case it would need to have all the iterator plumbing, etc.  Providing this capability is not too difficult, requiring only a fairly straightforward facade.
* cpos.h - Customization point objects to work in conjunction with the concepts, including
  * `num_vectors()`, `dimension()` to provide abstract size data for set of vectors (rather than `num_rows()`, `num_cols()`)
  * `load()` - abstracts away loading data into a matrix.  Is a nop for a `Matrix`, will load data into a `tdbMatrix`

Further descriptions about the CPP_API PR can be found in PR 131.

Graph-Based Indexing and Search
=========================

* Graph classes
  * adj_list.h - Two adjacency list classes, one that just stores indices and one that stores indices and scores
  * nn-graph.h - An adjacency list class `nn_graph` that keeps the neighbors in a `fixed_min_heap` in order to store only the neighbors with the top-n scores

Graph creation and search algorithms
* nn-descent.h - Implements the nn-descent algorithm, using an `nn_graph`
* vamana.h - Implements a `vamana_index` class that builds an index according to the vamana algorithm as described in the "filtered fresh DiskANN" paper (*not* either the original DiskANN paper nor the fresh DiskANN paper).  
  * The index is built by iterating over the vectors in the data set and processing each vector as an insert into the index.  
  * When a vector is added, a best-first path from the medoid (the vector closest to the middle of the vector data set) is computed. 
  *  The path is then pruned to keep the degree of the vertexes on that path below a given threshold.  
  * The functions for performing the best-first search and the pruning are `greedy_search` and `robust_prune` (which are the names given in the original DiskANN paper).  

Utility and helper functions
* diskann.h - functions for reading DiskANN memory index files

Demo applications
* The `demo_*` programs in the test subdirectory are essentially integration tests that output qualitative data for assessing proper functioning of key algorithms in building the graph-based indexes (such as dumping graph information to disk so that the graphs can be read and plotted by networkx).  Two notebooks are provided for examining data dumped by `unit_vamana` and `vamana`: `plot_unit_vamana.ipynb` and `plot_vamana.ipynb`.

Unit tests
* Unit tests for the new classes and functions are provided in include/test.

CLI drivers
* Two CLI drivers are provided for vamana indexing: src/vamana/index.cc which builds a vamana-based index for a specified TileDB set of vector data, and src/vamana/query.cc that queries a vamana-based index.  As with previous C++ CLI programs, these print extensive logging data.

Benchmark scripts
* src/benchmarks/vamana - diskann.bash and vamana.bash respectively run indexing and querying programs from DiskANN and TileDB on some small-ish datasets (datasets that will fit in memory) and report timings (etc).  DiskANN is only configured to compile and run on x86 Linux, so these scripts need to be run on an appropriate platform.

Cleanup of unused files

*NB*  A number of the unit tests read and write arrays from the local file system.  These are hard-coded to my personal development environment since we are not currently pulling those into a uniform location (this should be fixed at some point so that test arrays are pulled in manually by running a script or automatically as part of cmake).

### CODE REVIEW

**Comment by jparismorgan** [comment]
Is this PR meant to go before #161, which looks like a superset of this? Generally seems good, though would be good to apply my review on #161 here before landing.

Approving, though it would be good to get another review here as well.

**Review by jparismorgan — APPROVED** [review]

**Comment by lums658** [comment]
> Is this PR meant to go before #161, which looks like a superset of this? Generally seems good, though would be good to apply my review on #161 here before landing.
> 
> Approving, though it would be good to get another review here as well.

The idea here was to do a merge with the latest main so that there were no conflicts with main and so that it could be quickly approved and merged back into main.  From there, we can rebase #153, #154, #155, and #161 so that their diffs will not be so large and hopefully get them reviewed and approved relatively quickly so that we can make and keep main consistent with the latest goings on with C++.

File: apis/python/setup.py
Comment (NikolaosPapailiou) [inline apis/python/setup.py:37]: Remove

File: src/cmake/HighFive.cmake
Comment (NikolaosPapailiou) [inline src/cmake/HighFive.cmake:1]: Why is this needed?

File: src/include/detail/graph/diskann.h
Comment (NikolaosPapailiou) [inline src/include/detail/graph/diskann.h:47]: This should be TileDB Array, what is the plan to use TileDB storage for diskann?

File: src/include/detail/graph/nn-descent.h
Comment (NikolaosPapailiou) [inline src/include/detail/graph/nn-descent.h:168]: Remove commented lines here and in other parts of the file. If something need to remain add a comment for why it is needed.

File: src/include/detail/graph/nn-descent.h
Comment (NikolaosPapailiou) [inline src/include/detail/graph/nn-descent.h:181]: Remove dead code if not needed

File: src/include/detail/graph/nn-descent.h
Comment (NikolaosPapailiou) [inline src/include/detail/graph/nn-descent.h:274]: Having arguments specified as `auto` greatly reduces the readability of this code

File: src/include/detail/ivf/qv.h
Comment (NikolaosPapailiou) [inline src/include/detail/ivf/qv.h:1305]: Please remove lines that are commented out. Same bellow

File: src/include/detail/linalg/matrix.h
Comment (NikolaosPapailiou) [inline src/include/detail/linalg/matrix.h:2]: nit matrix

File: src/include/detail/linalg/matrix.h
Comment (NikolaosPapailiou) [inline src/include/detail/linalg/matrix.h:172]: Avoid dead code

File: src/include/test/time_abstraction_penalty.cc
Comment (NikolaosPapailiou) [inline src/include/test/time_abstraction_penalty.cc:1]: Remove if not needed

File: src/include/test/unit_high_five.cc
Comment (NikolaosPapailiou) [inline src/include/test/unit_high_five.cc:1]: Remove

**Review by NikolaosPapailiou — COMMENTED** [review]
This PR has a lot of features packaged in a single review cycle. 

In terms of release safety, we definitely need CI to be enabled and the test to pass. I would also prefer to split this in multiple PRs that contain the distinct feature functionalities (i.e. CPOs, concepts, Diskann, Vamana, C++ API)

**Comment by ihnorton** [comment]
#173+

**Pull request closed without merging**
